### DISCUSSION TITLE
Creating a Launcher

### DISCUSSION THREAD

**Discussion by RoyiAvital** [winpython/winpython#1079] [Q&A] [answered]
Hello,

How could one create a launcher like the `VS Code` launcher (The `VS Code.exe` file)?
I'd like to create another launcher with a different path.

Thanks.

**Comment by firai** [comment]
See the build script, which uses the NSIS build script: https://github.com/winpython/winpython/blob/d77fb5d96f669fa5d9aa9346b287c7821fe1f950/make.py#L975-L981 https://github.com/winpython/winpython/blob/d77fb5d96f669fa5d9aa9346b287c7821fe1f950/make.py#L591-L632
https://github.com/winpython/winpython/blob/53696948ca6c8c9f01870918ab35fd763a2f9567/portable/launcher_basic.nsi#L1-L44

**Reply by RoyiAvital** [reply]
Is there a way to replicate this without cloning the whole project? Even in a less general manner (Like NSIS only or something like that).

**Reply by firai** [reply]
Then just modify the NSIS script?

**Reply by RoyiAvital** [reply]
Any chance to guide me?

I guess I need:

1. What to change?
2. What other files are required (I see something like `Noshell.vbs`, `winvscode.bat`)?
3. How to run?

Once I get them all I will create a template and add it in Wiki so other will be able to use it.
It will be very useful for cases we want to have multiple profiles of VS Code / Spyder for a single installation.

**Reply by RoyiAvital** [reply]
@stonebig , @firai , Any chance for assistance on this?

**Comment by stonebig** [comment]
I didn't retry to create launchers without NSIS recently.
I may retry this summer

**Reply by RoyiAvital** [reply]
I am OK with NSIS, but can we have an NSIS template which is a standalone (Not activated by the code)?

For instance, how to create an NSIS launcher which launces VSCode32 in the `t` folder?
Can you share an NSIS code for that with the additional required files?

**Comment by stonebig** [comment]
It's in the source, look for create_launcher
It call nsis object that maybe we can put in the winpython

**Reply by RoyiAvital** [reply]
It was written above but I don't get it.

Any chance you point me in the NSIS file?

**Comment by firai** [comment]
The snippets I quoted above are self explanatory as to what parameters the Python script is feeding into the NSIS script. If you want to avoid the Python script, then you need to manually feed all of those parameters into the NSIS script.

Whether you need `Noshell.vbs` and `winvscode.bat` depend on whether you need the functionality in those scripts. Only you can answer that question based on your intended usage.

Questions about NSIS specifically would be better served by reading NSIS's documentation, going to NSIS's forum or going on the NSIS Discord.

Overall, you might get better answers if you ask specific questions after you actually try to do it, as opposed to a general request for a tutorial.

**Reply by RoyiAvital** [reply]
@firai , This is not a friendly comment.

Had I known what's going on, I would be more specific.

For example, I can't even find `Noshell.vbs` and `winvscode.bat` on this repository, how can I know if I need them?
I see them on my installed distribution, does the `NSIS` compile them or just create an `exe` which calls them?
By search I get that `wscript.exe` is probably used to run the `vbs` file.

If I get it right, at the end we're just execute the NSIS script you linked to with updated values for:

```nsis
!addincludedir "" 
!define COMMAND "" 
!define PARAMETERS "" 
!define WORKDIR "" 
Icon "" 
OutFile "" 
```

As I wrote, what I'd like to be able, for my example, a launcher for VS Code 32 Bit which will be on the folder `t\VSCode32`.

I am here to learn and then support by documenting everything I learned in the Wiki. But please be kind and guide me through it. The benefit will be for others as well.

**Comment by firai** [comment]
They're generated by the `make.py` that I referred to above.

https://github.com/winpython/winpython/blob/53696948ca6c8c9f01870918ab35fd763a2f9567/make.py#L1407-L1423
https://github.com/winpython/winpython/blob/53696948ca6c8c9f01870918ab35fd763a2f9567/make.py#L1962-L1978

**Reply by RoyiAvital** [reply]
OK, So basically what I'm after is how the `nsi` file look like just before it is compiled.
I guess if someone has built the distribution it may share it?

I am trying to go down the rabbit hole.

### Step I
Running the function `create_launcher()` with `name = VS Code.exe`, `icon = pathToIcon`, `command = wscript.exe`, `args = Noshell.vbs winvscode.bat`, `workdir = r'$EXEDIR\scripts` and `launcher = launcher_basic.nsi`.

The function creates the list of tuples:

```python
 data = [
            ('WINPYDIR', '$EXEDIR\%s' % self.python_name),
            ('WINPYVER', self.winpyver),
            ('COMMAND', 'wscript.exe'),
            ('PARAMETERS', 'Noshell.vbs winvscode.bat'),
            ('WORKDIR', '$EXEDIR\scripts'),
            ('Icon', icon.ico),
            ('OutFile', 'VS Code.exe'),
        ]
```

### Step II

Running the `build_nsis()` function with `srcname = 'launcher_basic.nsi`, `dstname = 'VS Code.nsi'` and `data = data`  (The list from previous step).

Basically this function edits the base `nsi` file by replacing variables according to the tuples in `data`.  
It adds to the list the tuple `('!addincludedir', osp.join(portable_dir, 'include'),)`. This probably makes the icon files and other resources available to the NSIS compiler.

It edits (If I got it right) the lines (Insert values inside the `""`):

```nsis
!addincludedir ""
!define COMMAND ""
!define PARAMETERS ""
!define WORKDIR ""
Icon ""
OutFile ""
```

So 

```nsis
!addincludedir "PathToResourcesFolder"
!define COMMAND "wscript.exe"
!define PARAMETERS "Noshell.vbs winvscode.bat"
!define WORKDIR "$EXEDIR\scripts"
Icon "PathToIcon"
OutFile "VS Code.exe"
```

Am I right? Did I miss something?

**Comment by RoyiAvital** [comment, marked as answer]
OK, So basically what I'm after is how the `nsi` file look like just before it is compiled.
I guess if someone has built the distribution it may share it?

I am trying to go down the rabbit hole.

### Step I
Running the function `create_launcher()` with `name = VS Code.exe`, `icon = pathToIcon`, `command = wscript.exe`, `args = Noshell.vbs winvscode.bat`, `workdir = r'$EXEDIR\scripts` and `launcher = launcher_basic.nsi`.

The function creates the list of tuples:

```python
 data = [
            ('WINPYDIR', '$EXEDIR\%s' % self.python_name),
            ('WINPYVER', self.winpyver),
            ('COMMAND', 'wscript.exe'),
            ('PARAMETERS', 'Noshell.vbs winvscode.bat'),
            ('WORKDIR', '$EXEDIR\scripts'),
            ('Icon', icon.ico),
            ('OutFile', 'VS Code.exe'),
        ]
```

### Step II

Running the `build_nsis()` function with `srcname = 'launcher_basic.nsi`, `dstname = 'VS Code.nsi'` and `data = data`  (The list from previous step).

Basically this function edits the base `nsi` file by replacing variables according to the tuples in `data`.  
It adds to the list the tuple `('!addincludedir', osp.join(portable_dir, 'include'),)`. This probably makes the icon files and other resources available to the NSIS compiler.

It edits (If I got it right) the lines (Insert values inside the `""`):

```nsis
!addincludedir ""
!define COMMAND ""
!define PARAMETERS ""
!define WORKDIR ""
Icon ""
OutFile ""
```

So 

```nsis
!addincludedir "PathToResourcesFolder"
!define COMMAND "wscript.exe"
!define PARAMETERS "Noshell.vbs winvscode.bat"
!define WORKDIR "$EXEDIR\scripts"
Icon "PathToIcon"
OutFile "VS Code.exe"
```

Am I right? Did I miss something?
